README.md
## Description

`tsvtree` is a command line tool to analyse TSV (Tab Separated Value) files. Its main functionalities are

* Visualization of the TSV content in a tree-like format.
* Compression and serialization.
* Node information: The number leaf nodes are reachable from an entry and its coordinate in the tree.
* Support for `TikZ` output format.

It can be seen as a complement to text processing tools like `awk`, `cut`, `column`, `tac` etc. An example `TikZ` image can be seen below.

![](http://db.occase.de/tsvtree.png)

## Showing the tree

Assume a TSV file like the one in `examples/cities.tsv` shown below

```bash
$ column examples/cities.tsv -t -s $'\t'
Earth  Europe   Western Europe  Germany  Baden-Württemberg    Karlsruhe     Karlsruhe                        Bruchsal
Earth  America  South America   Brazil   Sudeste              São Paulo     Macro Metropolitana Paulista     Bragança Paulista  Piracaia     Batatuba
Earth  America  South America   Brazil   Sudeste              São Paulo     Macro Metropolitana Paulista     Bragança Paulista  Atibaia      Alvinópolis
Earth  Europe   Western Europe  Germany  Baden-Württemberg    Karlsruhe     Rastatt                          Bühl
Earth  Europe   Western Europe  Germany  Nordrhein-Westfalen  Arnsberg      Ennepe-Ruhr-Kreis                Gevelsberg
Earth  America  South America   Brazil   Sudeste              São Paulo     Macro Metropolitana Paulista     Bragança Paulista  Atibaia      Jardim Siriema
Earth  Europe   Western Europe  Germany  Baden-Württemberg    Karlsruhe     Karlsruhe                        Karlsbad
Earth  America  South America   Brazil   Sudeste              São Paulo     Metropolitana de São Paulo       São Paulo          São Paulo    Mooca
Earth  America  South America   Brazil   Sudeste              Minas Gerais  Metropolitana de Belo Horizonte  Itabira            Alvinópolis  Alvinópolis
Earth  America  South America   Brazil   Sudeste              São Paulo     Macro Metropolitana Paulista     Bragança Paulista  Atibaia      Vila Santista
Earth  Europe   Western Europe  Germany  Nordrhein-Westfalen  Arnsberg      Ennepe-Ruhr-Kreis                Hattingen
Earth  Europe   Western Europe  France   Grand Est            Elsass        Bas-Rhin                         Strasbourg
Earth  America  South America   Brazil   Sudeste              Minas Gerais  Metropolitana de Belo Horizonte  Itabira            Alvinópolis  Major Ezequiel
Earth  Europe   Western Europe  Germany  Baden-Württemberg    Karlsruhe     Rastatt                          Bühlertal
Earth  Europe   Western Europe  France   Île-de-France        Val-de-Marne  Sucy-en-Brie
Earth  America  South America   Brazil   Sudeste              São Paulo     Metropolitana de São Paulo       São Paulo          São Paulo    Vila Leopoldina
Earth  Europe   Western Europe  France   Grand Est            Elsass        Bas-Rhin                         Haguenau
```
It is difficult to see the hierarchical nature of the data by looking at it in this form, also, the redundancy is distracting. `tsvtree` can display it in a much more comprehensible tree-like format as shown below

```bash
$ tsvtree examples/cities.tsv
Earth
├── America
│   └── South America
│       └── Brazil
│           └── Sudeste
│               ├── Minas Gerais
│               │   └── Metropolitana de Belo Horizonte
│               │       └── Itabira
│               │           └── Alvinópolis
│               │               ├── Alvinópolis
│               │               └── Major Ezequiel
│               └── São Paulo
│                   ├── Macro Metropolitana Paulista
│                   │   └── Bragança Paulista
│                   │       ├── Atibaia
│                   │       │   ├── Alvinópolis
│                   │       │   ├── Jardim Siriema
│                   │       │   └── Vila Santista
│                   │       └── Piracaia
│                   │           └── Batatuba
│                   └── Metropolitana de São Paulo
│                       └── São Paulo
│                           └── São Paulo
│                               ├── Mooca
│                               └── Vila Leopoldina
└── Europe
    └── Western Europe
        ├── France
        │   ├── Grand Est
        │   │   └── Elsass
        │   │       └── Bas-Rhin
        │   │           ├── Haguenau
        │   │           └── Strasbourg
        │   └── Île-de-France
        │       └── Val-de-Marne
        │           └── Sucy-en-Brie
        └── Germany
            ├── Baden-Württemberg
            │   └── Karlsruhe
            │       ├── Karlsruhe
            │       │   ├── Bruchsal
            │       │   └── Karlsbad
            │       └── Rastatt
            │           ├── Bühl
            │           └── Bühlertal
            └── Nordrhein-Westfalen
                └── Arnsberg
                    └── Ennepe-Ruhr-Kreis
                        ├── Gevelsberg
                        └── Hattingen
```
### Reading from stdin

To interact with other text processing tools like the ones mentioned above `tsvtree` can also read from standard input. Here are some illustrative examples.

* The authors of the last 50 commits on the linux kernel sorted by date
```bash
git log --oneline --pretty="%as%x09%an" -n 50 | tsvtree
Master
├── 2020-04-09
│   └── Masahiro Yamada
├── 2020-04-10
│   ├── Anshuman Khandual
│   ├── Arjun Roy
│   ├── Eric Biggers
│   ├── Jaewon Kim
│   ├── Linus Torvalds
│   ├── Logan Gunthorpe
│   ├── Pali Rohár
│   ├── Roman Gushchin
│   ├── Thomas Gleixner
│   ├── Vasily Averin
│   ├── Xiaoyao Li
│   └── kbuild test robot
├── 2020-04-11
│   ├── Linus Torvalds
│   ├── Sedat Dilek
│   └── Trond Myklebust
└── 2020-04-12
    └── Linus Torvalds
```
* Last 40 commits on the linux kernel excluding Linus Torvalds sorted by author.

```bash
git log --oneline --pretty="%ci%x09%h%x09%s%x09%an" -n 40 | grep -v Torvalds | awk -F $'\t' '{print $4"\t"$1" "$2" "$3}' | tsvtree
Master
├── Eric Biggers
│   ├── 2020-04-10 15:36:22 -0700 23756e551f35 selftests: kmod: test disabling module autoloading
│   ├── 2020-04-10 15:36:22 -0700 26c5d78c976c fs/filesystems.c: downgrade user-reachable WARN_ONCE() to pr_warn_once()
│   └── 2020-04-10 15:36:22 -0700 d7d27cfc5cf0 kmod: make request_module() return an error when autoloading is disabled
├── Logan Gunthorpe
│   ├── 2020-04-10 15:36:21 -0700 bfeb022f8fe4 mm/memory_hotplug: add pgprot_t to mhp_params
│   ├── 2020-04-10 15:36:21 -0700 c164fbb40c43 x86/mm: thread pgprot_t through init_memory_mapping()
│   └── 2020-04-10 15:36:21 -0700 f5637d3b42ab mm/memory_hotplug: rename mhp_restrictions to mhp_params
│   └── 2020-04-11 11:42:35 -0400 27d231c0c63b pNFS: Fix RCU lock leakage
├── Vasily Averin
│   ├── 2020-04-10 15:36:22 -0700 3bfa7e141b0b fs/seq_file.c: seq_read(): add info message about buggy .next functions
│   ├── 2020-04-10 15:36:22 -0700 89163f93c6f9 ipc/util.c: sysvipc_find_ipc() should increase position index
│   └── 2020-04-10 15:36:22 -0700 f4d74ef6220c kernel/gcov/fs.c: gcov_seq_next() should increase position index
└── Xiaoyao Li
    ├── 2020-04-11 16:40:55 +0200 9de6fe3c28d6 KVM: x86: Emulate split-lock access as a write in emulator
    └── 2020-04-11 16:42:41 +0200 e6f8b6c12f03 KVM: VMX: Extend VMXs #AC interceptor to handle split lock #AC in guest
```
* Last access time on each file of an arbitrary repository in th form `root:month:day`. 

```bash
$ find  . -name *.cpp -printf "%Tm\t%Td\t%f\n" | tsvtree
Root
├── 01
│   ├── 06
│   │   ├── redis.cpp
│   │   └── test_clients.cpp
│   └── 09
│       ├── notify-test.cpp
│       └── mms.cpp
├── 02
│   ├── 01
│   │   ├── db.cpp
│   │   └── key-gen.cpp
│   ├── 11
│   │   └── crypto.cpp
│   └── 12
│       ├── mms_session.cpp
│       └── post.cpp
├── 04
│   └── 06
│       └── json.cpp
└── 12
    ├── 18
    │   ├── db_tests.cpp
    │   ├── logger.cpp
    │   ├── net.cpp
    │   └── system.cpp
    ├── 28
    │   ├── notifier.cpp
    │   ├── ntf_session.cpp
    │   └── notify.cpp
    └── 31
        └── sim.cpp

```

### Restricting the depth

For large files it is sometimes useful to restrict the output to a certain depth in the tree. As an example let us run `tsvtree` on file [this](https://github.com/mzimbres/trees/blob/master/worldcities.tree.comp.tsv) file that contains all cities in the world restricting the output depth to 2

```bash
$ tsvtree --tree --depth 2  worldcities.comp
Places
├── Africa
│   ├── Northern Africa
│   ├── Eastern Africa
│   ├── Middle Africa
│   ├── Southern Africa
│   └── Western Africa
├── America
│   ├── Caribbean
│   ├── Central America
│   ├── South America
│   └── Northern America
├── Asia
│   ├── Central Asia
│   ├── Eastern Asia
│   ├── South-eastern Asia
│   ├── Southern Asia
│   └── Western Asia
├── Europe
│   ├── Northern Europe
│   ├── Eastern Europe
│   ├── Southern Europe
│   └── Western Europe
└── Oceania
    ├── Autralia and New Zealand
    ├── Melanesia
    ├── Micronesia
    └── Polynesia
```
The `.comp` format will be explained below.

## Compressing the tree

TSV files like the one above have a lot of redundancy since the nodes at small depths appear many times in the file. With `tsvtree` it is possible to compress the file by using numbers to represent the depth of the tree. The resulting file can be more than five times smaller then the original TSV

```bash
$ tsvtree --output comp examples/cities.tsv 
0	Root
1	Brazil
2	Rio de Janeiro
3	Rio de Janeiro
3	Teresópolis
2	Sao Paulo
3	Atibaia
3	Sao Paulo
1	France
2	Alsacia
3	Hagenau
3	Strasburg
2	Île-de-France
3	Paris
3	Sucy-en-Brie
1	Germany
2	Baden-würtemberg
3	Karlsruhe
3	Stuttgart
2	Nordrhein-Westphalen
3	Düsseldorf
3	Köln
```

The file `worldcities.comp` in the compressed format has the 456 kbytes in size whereas the same data in TSV format has 2064 kbytes, namely more that five times larger. To decompress it again to either TSV or tree format run

```bash
$ tsvtree --tree --output tsv worldcities.comp
$ tsvtree --tree --output tree worldcities.comp
```
Compressing the data in this form with `gzip` or other compressing tools also produces better results than compressing the original TSV directly. 

## Leaf counters and node coordinate

Once the tree in `.comp` format is available it is possible to show some useful information about the nodes at a specific depth, such as the number of leaf nodes that are reachable from each node in the tree and their coordinate in the tree. For example

```bash
$ tsvtree --tree --depth 2 --output info worldcities.comp | column -t -s $'\t'
Northern Africa           000.000  99
Eastern Africa            000.001  109
Middle Africa             000.002  70
Southern Africa           000.003  219
Western Africa            000.004  143
Caribbean                 001.000  102
Central America           001.001  226
South America             001.002  11444
Northern America          001.003  366
Central Asia              002.000  42
Eastern Asia              002.001  1358
South-eastern Asia        002.002  380
Southern Asia             002.003  437
Western Asia              002.004  206
Northern Europe           003.000  829
Eastern Europe            003.001  2039
Southern Europe           003.002  1215
Western Europe            003.003  46532
Autralia and New Zealand  004.000  111
Melanesia                 004.001  77
Micronesia                004.002  10
Polynesia                 004.003  14
```
If your TSV file does not have the root node on the left and the leaf nodes on the right, you will have to reorder it. This can be easily achieved with `awk`, for example

```bash
awk -F$'\t' '{print $3":"$2":"$1}' file.tsv
```
will select columns 1, 2, and 3 of `file.tsv` and reverse them.



debian/changelog
tsvtree (1.0.6-1) unstable; urgency=medium

  [ Marcelo Zimbres Silva ]
  * Ports the build system to autools.

 -- Marcelo Zimbres Silva <mzimbres@gmail.com>  Sat, 03 Apr 2021 11:15:41 +0200

tsvtree (1.0.5-1) unstable; urgency=medium

  [ Marcelo Zimbres Silva ]
  * New upstream version 1.0.5

 -- Marcelo Zimbres Silva <mzimbres@gmail.com>  Thu, 21 May 2020 12:52:41 +0200

tsvtree (1.0.3-1) unstable; urgency=medium

  * Removes unneeded files
  * First steps with the debian pipeline
  * Adds content to the pipeline
  * New upstream version 1.0.3

 -- Marcelo Zimbres Silva <mzimbres@gmail.com>  Sat, 25 Apr 2020 13:03:02 +0200

tsvtree (1.0.2-1) unstable; urgency=medium

  * New upstream release

 -- Marcelo Zimbres Silva <mzimbres@gmail.com>  Sun, 19 Apr 2020 16:53:49 +0200

tsvtree (1.0.1-1) stable; urgency=low

  * First release, closes: bug#956526.

 -- Marcelo Zimbres Silva <mzimbres@gmail.com>  Sun, 19 Apr 2020 12:31:00 +0200


src/tsvtree.cpp
/* Copyright (c) 2020 Marcelo Zimbres Silva (mzimbres at gmail dot com)
 *
 * This file is part of tsvtree.
 * 
 * tsvtree is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * tsvtree is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with tsvtree.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <stack>
#include <string>
#include <sstream>
#include <iostream>
#include <fstream>
#include <iterator>

#include <boost/program_options/options_description.hpp>
#include <boost/program_options/parsers.hpp>
#include <boost/program_options/variables_map.hpp>

#include "tsv.hpp"
#include "tree.hpp"
#include "utils.hpp"
#include "config.h"

namespace tsvtree {

namespace print
{

struct node_info {
   char sep = '\t';
   void operator()(tree::node const& node) const
   {
      std::cout
      << node.name << sep
      << to_string(node.code) << sep
      << node.leaf_counter << "\n";
   }
};

}

struct options {
   int oformat;
   int depth;

   char in_field_sep = '\t';
   char in_line_break = '\n';

   int out_indent = 5;
   char out_line_break = '\n';
   char out_field_sep = '\t';

   std::string file;
   bool check_leaf_min_depth = false;
   bool exit = false;
   bool tsv = true;
   bool decorate_tree = true;

   tree::config::tikz tikz_conf;

   auto
   make_tree_cfg(std::string const& content,
                 bool tsv_arg) const
   {
      auto const fmt =
         detect_iformat(content,
                        in_line_break,
                        in_field_sep,
                        tsv_arg);

      return tree::config
      { in_field_sep
      , in_line_break
      , fmt};
   }

   auto make_tsv_cfg() const
   {
      return tsv_cfg
      { out_indent
      , in_field_sep
      , out_field_sep
      , out_line_break
      , decorate_tree};
   }
};

auto
to_channels(tree_elem const& elem,
            tree::config const& cfg)
{
   tree t {elem.data, cfg};
   tree_level_view view {t, elem.depth};

   auto f = [](auto const& o)
      { return o.code; };

   std::vector<std::vector<int>> channels;
   std::transform(std::cbegin(view),
                  std::cend(view),
                  std::back_inserter(channels),
                  f);

   return channels;
}

auto readfile(std::string const& file)
{
   using iter_type = std::istreambuf_iterator<char>;

   if (std::empty(file))
      return std::string {iter_type {std::cin}, {}};

   std::ifstream ifs(file);
   return std::string {iter_type {ifs}, {}};
}

auto to_oformat(int i)
{
   if (i == 1) return tree::config::format::tabs;
   if (i == 2) return tree::config::format::counter;
   if (i == 7) return tree::config::format::deco;
   if (i == 8) return tree::config::format::tikz;

   throw std::runtime_error("to_oformat: Invalid input.");
   return tree::config::format::tabs;
}

struct tree_info {
   std::string file;
   int depth = 0;
   int version = 0;
};

/*  Converts the input in the form
 *
 *    file:depth:version
 *
 *  to the struct tree_info.
 */
tree_info decode_tree_info(std::string const& data, char sep = ':')
{
   if (std::count(std::cbegin(data), std::cend(data), sep) == 0)
      return {data, std::numeric_limits<int>::max(), 0};

   if (std::size(data) < 5)
      throw std::runtime_error("Invalid name length (n < 5).");

   if (data.front() == sep)
      throw std::runtime_error("Invalid tree info.");

   if (data.back() == sep)
      throw std::runtime_error("Invalid tree info.");

   auto const p1 = data.find_first_of(sep);

   if (p1 == std::string::npos)
      return {data, std::numeric_limits<int>::max(), 0};

   auto const p2 = data.find_first_of(sep, p1 + 1);

   if (p2 == p1 + 1)
      throw std::runtime_error("Invalid tree info.");

   if (p2 == std::string::npos)
      return { data.substr(0, p1)
             , std::stoi(data.substr(p1 + 1))
             , 0};
   
   return { data.substr(0, p1)
          , std::stoi(data.substr(p1 + 1))
          , std::stoi(data.substr(p2 + 1))};
}

auto to_tree_elem(options const& op)
{
   using iter_type = std::istreambuf_iterator<char>;

   auto const info = decode_tree_info(op.file);

   std::ifstream ifs(info.file);
   std::string tree_str {iter_type {ifs}, {}};

   auto const cfg = op.make_tree_cfg(tree_str, op.tsv);
   tree t {tree_str, cfg};

   auto const raw = 
      serialize(t,
                tree::config::format::counter,
                op.out_line_break,
                std::numeric_limits<int>::max(),
                op.out_field_sep,
		op.tikz_conf);

   return tree_elem {raw, info.depth, info.version};
}

auto op6(options const& op)
{
   auto const info = decode_tree_info(op.file);
   auto const str = readfile(info.file);
   tree t {str, op.make_tree_cfg(str, op.tsv)};

   tree_level_view view {t, op.depth};
   for (auto iter = std::begin(view); iter != std::end(view); ++iter) {
      auto const& line = iter.line();
      auto const ret = join(line, op.out_field_sep);
      if (!std::empty(ret))
         std::cout << ret << std::endl;
   }

   return 0;
}

auto op3(options const& op)
{
   auto const info = decode_tree_info(op.file);
   auto content = readfile(info.file);

   if (op.tsv) {
      auto const cfg = op.make_tsv_cfg();
      content = make_tree_string(content, cfg);
   }

   auto const cfg = op.make_tree_cfg(content, false);
   tree t {content, cfg};
   t.load_leaf_counters();

   tree_level_view view {t, op.depth};

   std::for_each(std::cbegin(view),
                 std::cend(view),
                 print::node_info{op.out_field_sep});

   return 0;
}

auto op1(options const& op)
{
   auto const info = decode_tree_info(op.file);
   auto const content = readfile(info.file);

   if (op.tsv) {
      auto const cfg = op.make_tsv_cfg();
      auto const out = make_tree_string(content, cfg);
      std::cout << out << std::flush;
      return 0;
   }

   auto const cfg = op.make_tree_cfg(content, op.tsv);
   tree t {content, cfg};
   auto const format = to_oformat(op.oformat);

   auto const out =
      serialize(t,
                format,
                op.out_line_break,
                op.depth,
                op.out_field_sep,
		op.tikz_conf);

   if (format == tree::config::format::tikz) {
      std::cout <<
      "\\documentclass[11pt]{article}\n"
      "\\usepackage{graphics}\n"
      "\\usepackage[dvipsnames]{xcolor}\n"
      "\\usepackage{tikz}\n"
      "\\usetikzlibrary{positioning}\n"
      "\\usetikzlibrary{arrows}\n"
      "\n"
      "\\colorlet{memC}{Apricot}\n"
      "\\colorlet{arrowC}{black!70!white}\n"
      "\\colorlet{docC}{RoyalBlue!50}\n"
      "\\colorlet{textC}{black!80}\n"
      "\\tikzstyle{treenode}=[anchor=south west, rounded corners=2pt, node distance=0pt, fill=memC,shape=rectangle,minimum height=12pt, minimum width=0pt, inner sep=3pt]\n"
      "\\tikzstyle{textnode}=[anchor=south west, rounded corners=2pt, node distance=0pt, shape=rectangle,minimum height=0.0cm  ,minimum width=0.5cm  ,inner sep=2pt]\n"
      "\\tikzstyle{marrow}=[very thick, densely dotted,>=stealth,->, color=black]\n"
      "\\tikzstyle{treearrow}=[rounded corners=8pt, very thick, >=stealth,<-, color=arrowC]\n"
      "\\def\\treenode{\\node[style=treenode]}\n"
      "\\def\\textnode{\\node[style=textnode]}\n"
      "\\def\\marrow{\\draw[style=marrow]}\n"
      "\\def\\treearrow{\\draw[style=treearrow]}\n"
      "\n"
      "\\pgfrealjobname{tree}\n"
      "\n"
      "\\begin{document}\n"
      "\\beginpgfgraphicnamed{tree-f0}\n"
      "   \\begin{tikzpicture}[scale=1.0]\n"
      "\\fill[color=docC] (-1,4) rectangle (24, -33);\n"
      "%\\shade[left color=BlueViolet!50,right color=BlueViolet!10] (-1,3) rectangle +(15,-15);\n"
      "\\textnode at (4, 3) {\\huge\\bf tsvtree};\n";
   }

   std::cout << out << std::flush;

   if (format == tree::config::format::tikz) {
      std::cout <<
      "\\end{tikzpicture}\n"
      "\\endpgfgraphicnamed\n"
      "\\end{document}\n";
   }

   return 0;
}

int check_leaf_min_depth_op(options const& op)
{
   auto const info = decode_tree_info(op.file);
   auto const str = readfile(info.file);
   auto const cfg = op.make_tree_cfg(str, op.tsv);

   tree t {str, cfg};

   auto const line = check_leaf_min_depths(t, op.depth);
   auto const out = join(line, op.out_field_sep);
   if (std::empty(out)) {
      std::cout << "Ok" << std::endl;
      return 0;
   }

   std::cout << "Error on line: " << out << std::endl;
   return 1;
}

int impl(options const& op)
{
   if (op.check_leaf_min_depth)
      return check_leaf_min_depth_op(op); 

   if (op.oformat == 1) return op1(op);
   if (op.oformat == 2) return op1(op);
   if (op.oformat == 3) return op3(op);
   if (op.oformat == 6) return op6(op);
   if (op.oformat == 7) return op1(op);
   if (op.oformat == 8) return op1(op);

   return 1;
}

}

namespace po = boost::program_options;
using namespace tsvtree;

auto make_oformat(std::string const& s, bool deco)
{
   if (s == "tree" && deco)     return 7;
   if (s == "tree")             return 1;
   if (s == "comp")             return 2;
   if (s == "info")             return 3;
   if (s == "tsv")              return 6;
   if (s == "tikz")             return 8;
   return -1;
}

auto parse_options(int argc, char* argv[])
{
   options op;
   std::string oformat = "tree";
   po::options_description desc("Options");
   desc.add_options()
   ( "help,h", "This help message.")
   ( "version,v", "Program version.")
   ( "tree,k", "Input file in tsv format.")
   ( "indent-with-tab,p", "Uses tab to represent the tree depth.")
   ( "depth,d", po::value<int>(&op.depth)->default_value(std::numeric_limits<int>::max()), "Influences the output.")
   ( "file,f", po::value<std::string>(&op.file), "The file containing the tree.")
   ( "input-separator,e", po::value<char>(&op.in_field_sep), "Field separator in the input file.")
   ( "input-line-break,r", po::value<char>(&op.in_line_break), "Line break in the input file.")
   ( "output-line-break,b", po::value<char>(&op.out_line_break), "Line break character used in the output.")
   ( "output-separator,s", po::value<char>(&op.out_field_sep), "Output field separator.")
   ( "check-leaf-min-depth,c","Checks whether all leaf nodes have at least the depth specified in --depth.")
   ( "output,o"
   , po::value<std::string>(&oformat)->default_value("tree")
   , "Format used in the output file. Available options:\n"
     "• tree: \tNode depth with indentation.\n"
     "• comp: \tCompressed tree.\n"
     "• info: \tInfo of nodes at --depth.\n"
     "• tsv:  \tTSV format.\n"
     "• tikz:  \tTikZ format."
   )
   ( "tikz-x-step,x", po::value<int>(&op.tikz_conf.x_step)->default_value(30), "Node horizontal distance in point units.")
   ( "tikz-y-step,y", po::value<int>(&op.tikz_conf.y_step)->default_value(20), "Node vertical distance in point units.")
   ( "tikz-color-min,a", po::value<int>(&op.tikz_conf.min)->default_value(30), "Right color minimum.")
   ( "tikz-color-max,g", po::value<int>(&op.tikz_conf.max)->default_value(60), "Right colot maximum.")
   ( "tikz-right-color,i", po::value<std::string>(&op.tikz_conf.right_color)->default_value("BurntOrange"), "Color of the right nodes.")
   ( "tikz-left-color,j", po::value<std::string>(&op.tikz_conf.left_color)->default_value("White"), "Color of the node left nodes.")
   ;

   po::positional_options_description pos;
   pos.add("file", -1);

   po::variables_map vm;        
   po::store(po::command_line_parser(argc, argv).
      options(desc).positional(pos).run(), vm);
   po::notify(vm);    

   op.decorate_tree = vm.count("indent-with-tab") == 0;
   op.oformat = make_oformat(oformat, op.decorate_tree);

   if (op.oformat == -1) {
      std::cerr << "Invalid output option." << std::endl;
      op.exit = true;
      return op;
   }

   if (vm.count("help")) {
      op.exit = true;
      std::cout << desc << "\n";
      return op;
   }

   if (vm.count("version")) {
      op.exit = true;
      std::cout << PACKAGE_VERSION << "\n";
      return op;
   }

   op.check_leaf_min_depth = vm.count("check-leaf-min-depth") > 1;
   op.tsv = vm.count("tree") == 0;

   if (op.tsv) {
      if (op.oformat == 2) op.out_indent = -1;
      if (op.oformat == 3) op.out_indent = -1;
   }

   return op;
}

int main(int argc, char* argv[])
{
   try {
      auto const op = parse_options(argc, argv);
      if (op.exit)
         return 0;

      return impl(op);
   } catch (std::exception const& e) {
      std::cerr << e.what() << std::endl;
      return 1;
   }
}

